library/src/main/scala/scala/scalajs/js/WeakRef.scala

```diff
@@ -0,0 +1,33 @@
+/*
+ * Scala.js (https://www.scala-js.org/)
+ *
+ * Copyright EPFL.
+ *
+ * Licensed under Apache License 2.0
+ * (https://www.apache.org/licenses/LICENSE-2.0).
+ *
+ * See the NOTICE file distributed with this work for
+ * additional information regarding copyright ownership.
+ */
+
+package scala.scalajs.js
+
+import scala.scalajs.js
+import scala.scalajs.js.annotation.JSGlobal
+
+/** <span class="badge badge-ecma2021" style="float: right;">ECMAScript 2021</span>
+ *
+ *  A WeakRef object lets you hold a weak reference to another object, without preventing
+ *  that object from getting garbage-collected.
+ *
+ *  MDN
+ *
+ *  @param targetObject An object whose weak reference contained by an instance of WeakRef.
+ *  @tparam T unconstrained, but the constructor will throw a TypeError if targetObject is not an Object.
+ */
+@js.native
+@JSGlobal
+class WeakRef[+T](targetObject: T) extends js.Object {
```

Comment:
Erf. Apparently Scala 2.11 does not like that `+T` here, although 2.12 and 2.13 are fine with it. Let's make it invariant for now.

Otherwise LGTM. We'll need to wait for #4458 to be merged, for the version bump (it's a new API so it needs a minor release), but otherwise it's good to go.

Reply:
Is this fix https://github.com/scala-js/scala-js/pull/4366/commits/51c9b9ad0fe6cf51594bc3b7daca476ee7225e26 fine ?

Reply:
Good question. I think it's fine, but we should test that calling `deref()` without type argument nor expected type infers the correct type. We typically do that in two steps, like this:
```scala
  @Test def testDeref(): Unit = {
    val obj = new js.Date()
    val ref = new js.WeakRef(obj)
    val derefedInferredType = ref.deref()
    val derefedAssertType: js.UndefOr[js.Date] = derefedInferredType
    assertTrue(derefedAssertType.contains(obj))
  }
```

---

Btw, #4458 has now been merged, so this PR can be rebased on master.